src/state_machine.zig

```diff
@@ -533,8 +533,38 @@ pub fn StateMachineType(
             callback(self);
         }
 
+        pub fn input_valid(operation: Operation, input: []align(16) const u8) bool {
+            switch (operation) {
+                .pulse => {
+                    if (input.len != 0) return false;
+                },
+                inline .get_account_transfers, .get_account_balances => |comptime_operation| {
+                    const event_size = @sizeOf(Event(comptime_operation));
+
+                    if (input.len != event_size) return false;
+                },
+                inline else => |comptime_operation| {
+                    const event_size = @sizeOf(Event(comptime_operation));
+                    const batch_max = comptime constants.batch_max.operation_batch_max(
+                        comptime_operation,
+                    );
+
+                    comptime assert(event_size > 0);
+                    comptime assert(batch_max > 0);
+
+                    if (input.len == 0) return false;
+                    if (input.len % event_size != 0) return false;
+                    if (input.len > batch_max * event_size) return false;
+                },
+            }
+
+            return true;
+        }
+
         /// Updates `prepare_timestamp` to the highest timestamp of the response.
-        pub fn prepare(self: *StateMachine, operation: Operation, input: []align(16) u8) void {
+        pub fn prepare(self: *StateMachine, operation: Operation, input: []align(16) const u8) void {
```

Comment:
Is there a reason for the input to be `input: []align(16) u8` instead of `input: []align(16) const u8` ?

Reply:
IIRC: Even though our state machine does not need to modify the input, it is in principle safe for the replica to do so. But other state machines might need/want to modify a request when it arrives.

Reply:
The way comptime "interfaces" work, each state machine can define whether it requires a mutable slice or not, without changing the contract with the replica.